pkg/bloomcompactor/bloomcompactor.go

```diff
@@ -158,26 +168,288 @@ func New(cfg Config,
+		}
+
+		// Ensure the tenant ID belongs to our shard.
+		owned, err := c.sharding.OwnsTenant(tenant)
+		if err != nil {
+			c.metrics.compactionRunSkippedTenants.Inc()
+			level.Warn(c.logger).Log("msg", "unable to check if tenant is owned by this shard", "tenantID", tenant, "err", err)
+			continue
+		}
+		if !owned {
+			c.metrics.compactionRunSkippedTenants.Inc()
+			level.Debug(c.logger).Log("msg", "skipping tenant because it is not owned by this shard", "tenantID", tenant)
+			continue
+		}
+
+		ownedTenants[tenant] = struct{}{}
+
+		if err := c.compactTenantWithRetries(ctx, sc, tableName, tenant); err != nil {
+			switch {
+			case errors.Is(err, context.Canceled):
+				// We don't want to count shutdowns as failed compactions because we will pick up with the rest of the compaction after the restart.
+				level.Info(c.logger).Log("msg", "compaction for tenant was interrupted by a shutdown", "tenant", tenant)
+				return nil
+			default:
+				c.metrics.compactionRunFailedTenants.Inc()
+				level.Error(c.logger).Log("msg", "failed to compact tenant", "tenant", tenant, "err", err)
+				errs.Add(err)
+			}
+			continue
+		}
+
+		c.metrics.compactionRunSucceededTenants.Inc()
+		level.Info(c.logger).Log("msg", "successfully compacted tenant", "tenant", tenant)
+	}
+
+	return errs.Err()
+
+	// TODO: Delete local files for unowned tenants, if there are any.
+}
+
+func (c *Compactor) compactTenant(ctx context.Context, sc storeClient, tableName string, tenant string) error {
+	level.Info(c.logger).Log("msg", "starting compaction of tenant", "tenant", tenant)
+
+	// Ensure the context has not been canceled (ie. compactor shutdown has been triggered).
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	// TODO: Use ForEachConcurrent?
+	errs := multierror.New()
+	if err := sc.indexShipper.ForEach(ctx, tableName, tenant, func(isMultiTenantIndex bool, idx shipperindex.Index) error {
+		if isMultiTenantIndex {
+			return fmt.Errorf("unexpected multi-tenant")
+		}
+
+		// TODO: Make these casts safely
+		if err := idx.(*tsdb.TSDBFile).Index.(*tsdb.TSDBIndex).ForSeries(
+			ctx, nil,
+			0, math.MaxInt64, // TODO: Replace with MaxLookBackPeriod
+			func(labels labels.Labels, fingerprint model.Fingerprint, chksMetas []tsdbindex.ChunkMeta) {
+				job := NewJob(tenant, tableName, fingerprint, chksMetas)
+
+				ownsJob, err := c.sharding.OwnsJob(job)
+				if err != nil {
+					c.metrics.compactionRunSkippedJobs.Inc()
+					level.Error(c.logger).Log("msg", "failed to check if compactor owns job", "job", job, "err", err)
+					errs.Add(err)
+					return
+				}
+				if !ownsJob {
+					c.metrics.compactionRunSkippedJobs.Inc()
+					level.Debug(c.logger).Log("msg", "skipping job because it is not owned by this shard", "job", job)
+					return
+				}
+
+				if err := c.runCompact(ctx, c.bloomShipperClient, sc, job); err != nil {
+					c.metrics.compactionRunFailedJobs.Inc()
+					errs.Add(errors.Wrap(err, "runBloomCompact"))
+					return
+				}
```

Comment:
Probably a personal preference, but I would separate job collection and job execution into separate functions.

Reply:
Agree! Let's do it on a followup PR once we add more logic to the actual compaction of chunks. Maybe moving that logic to `chunkcompactor.go`